File: template/app/src/main/java/com/example/base/baseapp/utils/AbstractRecyclerViewAdapter.java
package com.example.base.baseapp.utils;

import android.support.annotation.NonNull;
import android.support.v7.util.SortedList;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import butterknife.ButterKnife;
import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;

public abstract class AbstractRecyclerViewAdapter<T, E extends AbstractRecyclerViewAdapter.ItemHolder> extends RecyclerView.Adapter<AbstractRecyclerViewAdapter.ItemHolder> {

    private final PublishSubject<T> onItemClickSubject = PublishSubject.create();
    private final PublishSubject<T> onItemLongClickSubject = PublishSubject.create();
    private final PublishSubject<View> onConstantClickSubject = PublishSubject.create();

    protected final SortedList<T> elements;
    private final ConstantElement additionalElement;
    private List<T> originalElements;

    public enum ViewType {
        ITEM, CONSTANT_ITEM
    }

    public enum ConstantElement {
        NONE, TOP, BOTTOM
    }

    public AbstractRecyclerViewAdapter(Class<T> tClass) {
        this(tClass, ConstantElement.NONE);
    }

    public AbstractRecyclerViewAdapter(Class<T> tClass, ConstantElement constantElement) {
        this.additionalElement = constantElement;
        elements = new SortedList<>(tClass, new SortedList.Callback<T>() {

            @Override
            public void onInserted(int position, int count) {
                notifyItemRangeInserted(position, count);
            }

            @Override
            public void onRemoved(int position, int count) {
                notifyItemRangeRemoved(position, count);
            }

            @Override
            public void onMoved(int fromPosition, int toPosition) {
                notifyItemMoved(fromPosition, toPosition);
            }

            @Override
            public int compare(T o1, T o2) {
                return compareElements(o1, o2);
            }

            @Override
            public void onChanged(int position, int count) {
                notifyItemRangeChanged(position, count);
            }

            @Override
            public boolean areContentsTheSame(T oldItem, T newItem) {
                return oldItem.equals(newItem);
            }

            @Override
            public boolean areItemsTheSame(T item1, T item2) {
                return item1 == item2;
            }
        });
    }

    @NonNull
    @Override
    public ItemHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewTypeInt) {
        ViewType viewType = ViewType.values()[viewTypeInt];
        View view;
        if (viewType == ViewType.ITEM) {
            view = LayoutInflater.from(parent.getContext())
                    .inflate(getViewHolderLayoutResource(), parent, false);
            return instantiateViewHolder(view);
        } else {
            view = LayoutInflater.from(parent.getContext())
                    .inflate(getConstantItemLayoutResource(), parent, false);
            return new ItemHolder(view);
        }
    }

    protected int getConstantItemLayoutResource() {
        throw new UnsupportedOperationException("Implement getConstantItemLayoutResource in your adapter");
    }

    @Override
    public void onBindViewHolder(@NonNull AbstractRecyclerViewAdapter.ItemHolder holder, int position) {
        if (position < elements.size()) {
            T element = elements.get(position);
            holder.itemView.setOnClickListener(v -> onItemClickSubject.onNext(element));
            holder.itemView.setOnLongClickListener(v -> {
                onItemLongClickSubject.onNext(element);
                return true;
            });
            onBindViewHolder((E) holder, element);
        } else {
            holder.itemView.setOnClickListener(v -> onConstantClickSubject.onNext(holder.itemView));
        }
    }

    @Override
    public int getItemViewType(int position) {
        switch (additionalElement) {
            case TOP:
                return position == 0 ? ViewType.CONSTANT_ITEM.ordinal() : ViewType.ITEM.ordinal();
            case BOTTOM:
                return position == elements.size() ? ViewType.CONSTANT_ITEM.ordinal() : ViewType.ITEM.ordinal();
            default:
                return ViewType.ITEM.ordinal();
        }
    }

    @Override
    public int getItemCount() {
        return elements.size() + (additionalElement != ConstantElement.NONE ? 1 : 0);
    }

    public void setElements(List<T> newElements) {
        originalElements = new ArrayList<>(newElements);
        elements.replaceAll(originalElements);
    }

    public Observable<T> getElementClicks() {
        return onItemClickSubject.hide();
    }

    public Observable<T> getElementLongClicks() {
        return onItemLongClickSubject.hide();
    }

    public Observable<View> getConstantElementClicks() {
        return onConstantClickSubject.hide();
    }

    protected int compareElements(T o1, T o2) {
        return originalElements.indexOf(o1) - originalElements.indexOf(o2);
    }

    protected abstract E instantiateViewHolder(View view);

    protected abstract int getViewHolderLayoutResource();

    protected abstract void onBindViewHolder(E holder, T t);

    public static class ItemHolder extends RecyclerView.ViewHolder {
        public ItemHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: generators/app/templates/app/src/main/kotlin/com/example/base/baseapp/utils/AbstractRecyclerViewAdapter.kt
package com.example.base.baseapp.utils

import android.support.v7.util.SortedList
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import java.util.ArrayList

import butterknife.ButterKnife
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject

abstract class AbstractRecyclerViewAdapter<T, E : AbstractRecyclerViewAdapter.ItemHolder> (tClass: Class<T>, private val additionalElement: ConstantElement = ConstantElement.NONE) : RecyclerView.Adapter<AbstractRecyclerViewAdapter.ItemHolder>() {

    private val onItemClickSubject = PublishSubject.create<T>()
    private val onItemLongClickSubject = PublishSubject.create<T>()
    private val onConstantClickSubject = PublishSubject.create<View>()

    protected val elements: SortedList<T>
    private var originalElements: List<T>? = null

    protected val constantItemLayoutResource: Int
        get() =
            throw UnsupportedOperationException("Implement getConstantItemLayoutResource in your adapter")

    val elementClicks: Observable<T>
        get() = onItemClickSubject.hide()

    val elementLongClicks: Observable<T>
        get() = onItemLongClickSubject.hide()

    val constantElementClicks: Observable<View>
        get() = onConstantClickSubject.hide()

    protected abstract val viewHolderLayoutResource: Int

    enum class ViewType {
        ITEM, CONSTANT_ITEM
    }

    enum class ConstantElement {
        NONE, TOP, BOTTOM
    }

    init {
        elements = SortedList(tClass, object : SortedList.Callback<T>() {

            override fun onInserted(position: Int, count: Int) {
                notifyItemRangeInserted(position, count)
            }

            override fun onRemoved(position: Int, count: Int) {
                notifyItemRangeRemoved(position, count)
            }

            override fun onMoved(fromPosition: Int, toPosition: Int) {
                notifyItemMoved(fromPosition, toPosition)
            }

            override fun compare(o1: T, o2: T): Int {
                return compareElements(o1, o2)
            }

            override fun onChanged(position: Int, count: Int) {
                notifyItemRangeChanged(position, count)
            }

            override fun areContentsTheSame(oldItem: T, newItem: T): Boolean {
                return oldItem == newItem
            }

            override fun areItemsTheSame(item1: T, item2: T): Boolean {
                return item1 === item2
            }
        })
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewTypeInt: Int): ItemHolder {
        val viewType = ViewType.values()[viewTypeInt]
        val view: View
        if (viewType == ViewType.ITEM) {
            view = LayoutInflater.from(parent.context)
                    .inflate(viewHolderLayoutResource, parent, false)
            return instantiateViewHolder(view)
        } else {
            view = LayoutInflater.from(parent.context)
                    .inflate(constantItemLayoutResource, parent, false)
            return ItemHolder(view)
        }
    }

    override fun onBindViewHolder(holder: AbstractRecyclerViewAdapter.ItemHolder, position: Int) {
        if (position < elements.size()) {
            val element = elements.get(position)
            holder.itemView.setOnClickListener { _ -> onItemClickSubject.onNext(element) }
            holder.itemView.setOnLongClickListener { _ ->
                onItemLongClickSubject.onNext(element)
                true
            }
            @Suppress("UNCHECKED_CAST")
            onBindViewHolder(holder as E, element)
        } else {
            holder.itemView.setOnClickListener { _ -> onConstantClickSubject.onNext(holder.itemView) }
        }
    }

    override fun getItemViewType(position: Int): Int {
        when (additionalElement) {
            AbstractRecyclerViewAdapter.ConstantElement.TOP -> return if (position == 0) ViewType.CONSTANT_ITEM.ordinal else ViewType.ITEM.ordinal
            AbstractRecyclerViewAdapter.ConstantElement.BOTTOM -> return if (position == elements.size()) ViewType.CONSTANT_ITEM.ordinal else ViewType.ITEM.ordinal
            else -> return ViewType.ITEM.ordinal
        }
    }

    override fun getItemCount(): Int {
        return elements.size() + if (additionalElement != ConstantElement.NONE) 1 else 0
    }

    fun setElements(newElements: List<T>) {
        originalElements = ArrayList(newElements)
        elements.replaceAll(originalElements!!)
    }

    protected fun compareElements(o1: T, o2: T): Int {
        return originalElements!!.indexOf(o1) - originalElements!!.indexOf(o2)
    }

    protected abstract fun instantiateViewHolder(view: View): E

    protected abstract fun onBindViewHolder(holder: E, data: T)

    open class ItemHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        init {
            ButterKnife.bind(this, itemView)
        }
    }
}
